mimic/test/behavior_tests.py

```diff
@@ -0,0 +1,213 @@
+            trigger - this will be one of the ones specified in
+            :ivar:`name_and_params`.
+        :param response: The response from triggering the event.
+        :param body: The response body from triggering the event.
+        """
+
+    def validate_default_behavior(response, body):
+        """
+        Validate that the default behavior was triggered as opposed to
+        some other behavior.
+
+        :param response: The response from triggering the event.
+        :param body: The response body from triggering the event.
+        """
+
+    def from_test_case(test_case):
+        """
+        A constructor that generates a provider of
+        :class:`IBehaviorAPITestHelper`.
+
+        :param test_case: An instance of a test case that has assertion
+            and cleanup functions.
+        :return: a :class:`IBehaviorAPITestHelper` provider
+        """
+
+
+def register_behavior(test_case, root, uri, behavior_name, parameters,
+                      criteria):
+    """
+    Register a particular behavior.
+
+    :param test_case: the test case with which to make assertions
+    :param root: A mimic root API object
+    :param str uri: The uri fo the behavior resource to register.
+    :param str behavior_name: The name of the behavior
+    :param dict parameters: A dictionary of parameters to pass to the behavior
+    :param list criteria: The criteria for which this behavior should be
+        applied.
+
+    :return: The behavior ID of the registered behavior.
+    """
+    behavior_json = {"name": behavior_name,
+                     "parameters": parameters,
+                     "criteria": criteria}
+    response, body = test_case.successResultOf(json_request(
+        test_case, root, "POST", uri, json.dumps(behavior_json)))
+
+    test_case.assertEqual(response.code, 201)
+    behavior_id = body.get("id")
+    test_case.assertIsInstance(behavior_id, string_types)
+    test_case.assertEqual(UUID(behavior_id).version, 4)
+    return behavior_id
+
+
+def make_behavior_tests(behavior_helper_klass):
+    """
+    Will generate a test suite containing:
+
+    - 1 test case that deleting a behavior will revert to the underlying
+        behavior.  If 2 behaviors are provided, will also assert that the
+        first behavior registered will be the behavior used, and that deleting
+        it means the second mock takes over.  Deleting that will revert to
+        default behavior.  If only 1 behavior is provided, it will just test
+        that deleting it reverts to default behavior.
+
+    - 1 test case validating that deleting an invalid behavior for will
+      result in a 404.
+
+    - 1 test case asserting that providing invalid JSON will result in a 400
+      when creating the behavior.
+
+    :param behavior_helper_klass: a class that implements
+        :class:`IBehaviorAPITestHelper`
+
+    :return: a :class:`twisted.trial.unittest.SynchronousTestCase` object
```

Comment:
The type notation for the thing that is returned here would be a `SynchronousTestCase` *subclass*; saying an "object" vaguely implies an instance.